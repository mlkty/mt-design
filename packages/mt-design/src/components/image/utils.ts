
export const isValidImage = (src: string) => {
    return new Promise(resolve => {
        const image = new Image();
        image.onload = () => resolve(true);
        image.onerror = () => resolve(false);
        image.src = src;
    });
};
