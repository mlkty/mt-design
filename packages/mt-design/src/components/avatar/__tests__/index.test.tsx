import {UserOutlined} from '@ant-design/icons';

import {Avatar} from '..';
import {customRender, screen} from '../../../../tests/utils';
import mountTest from '../../../../tests/mountTest';

const LOAD_SUCCESS_SRC = 'LOAD_SUCCESS_SRC';

describe('Avatar', () => {
    mountTest(Avatar);

    beforeAll(() => {
        let src = '';
        Object.defineProperty(global.Image.prototype, 'src', {
            get() {
                return src;
            },
            set(_src) {
                src = _src;
                if (src.includes(LOAD_SUCCESS_SRC)) {
                    setTimeout(() => this.onload(), 200);
                } else {
                    setTimeout(() => this.onerror?.(new Error()), 200);
                }
            },
        });
    });

    it('render avatar by empty src', async () => {
        const {container} = customRender(<Avatar />);
        expect(container).toMatchSnapshot();
    });

    it('render avatar by src', async () => {
        const {container} = customRender(<Avatar src="LOAD_SUCCESS_SRC" />);
        expect(container).toMatchSnapshot();
    });

    it('render avatar by icon', async () => {
        const {container} = customRender(<Avatar icon={<UserOutlined />} />);
        expect(container.querySelector('svg')).toBeInTheDocument();
        expect(container.querySelector('img')).not.toBeInTheDocument();
    });

    it('render avatar by text', async () => {
        const {container} = customRender(<Avatar>U</Avatar>);
        expect(container.firstChild).toHaveTextContent('U');
    });

    // =========== style ==========
    const map = {
        small: 18,
        middle: 24,
        large: 32,
        '64': 64,
    };

    Object.entries(map).forEach(([size, value]: [any, number]) => {
        it(`render avatar size: ${size}`, async () => {
            const props: Record<string, any> = {};
            if (size !== 'middle') {
                props[size] = size;
            }
            customRender(<Avatar size={size}>U</Avatar>);
            expect(screen.getByRole('avatar')).toHaveStyle({fontSize: value});
        });
    });
});
