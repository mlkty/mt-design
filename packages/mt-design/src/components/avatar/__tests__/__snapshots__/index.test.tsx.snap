// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar render avatar by empty src 1`] = `
<div>
  <div
    class="mt-image mt-avatar mt-avatar--circle"
    role="avatar"
    style="font-size: 18px; width: 32px; height: 32px;"
  >
    <img
      class="mt-image-img"
      height="32"
      src=""
      style="object-fit: contain; height: 32px;"
      width="32"
    />
  </div>
</div>
`;

exports[`Avatar render avatar by src 1`] = `
<div>
  <div
    class="mt-image mt-avatar mt-avatar--circle"
    role="avatar"
    style="font-size: 18px; width: 32px; height: 32px;"
  >
    <img
      class="mt-image-img"
      height="32"
      src=""
      style="object-fit: contain; height: 32px;"
      width="32"
    />
    <div
      class="mt-image-intermediate"
      role="placeholder"
    />
  </div>
</div>
`;
